File: python/mirage/commands/builtin/generic/gunzip.py
from collections.abc import Awaitable, Callable, Mapping
from dataclasses import dataclass

from mirage.commands.builtin.utils.constants import STDIN_OPERAND
from mirage.commands.builtin.utils.stream import operand_label, stdin_bytes
from mirage.commands.config import CommandOpts
from mirage.commands.spec import SPECS
from mirage.commands.spec.flag_view import FlagView
from mirage.commands.spec.types import FlagValue
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec
from mirage.utils.compress import gunzip_checked
from mirage.utils.errors import GzipDataError
from mirage.utils.key_prefix import mounted_path


async def gunzip(
    paths: list[PathSpec],
    *,
    read_bytes: Callable[..., Awaitable[bytes]],
    write_bytes: Callable[..., Awaitable[None]],
    unlink: Callable[..., Awaitable[None]],
    stdin: ByteSource | None = None,
    keep: bool = False,
    force: bool = False,
    to_stdout: bool = False,
    test_only: bool = False,
) -> tuple[ByteSource | None, IOResult]:
    read = stdin_bytes(read_bytes, stdin)
    writes: dict[str, ByteSource] = {}
    stdout: list[bytes] = []
    errors: list[str] = []
    # With no operand gunzip reads stdin. A `-` has no file to replace, so
    # it decompresses to stdout; gzip refuses to follow /dev/stdin in
    # place, so only `-` does this. An input with no gzip header is
    # reported and skipped; a truncated or corrupt one ends the run.
    for p in paths or [STDIN_OPERAND]:
        in_place = not (to_stdout or test_only or p.raw_path == "-")
        raw = await (read_bytes(p) if in_place else read(p))
        try:
            data = gunzip_checked(raw)
        except GzipDataError as exc:
            errors.append(f"gunzip: {operand_label(p, 'stdin')}: {exc}\n")
            if exc.fatal:
                break
            continue
        if test_only:
            continue
        if not in_place:
            stdout.append(data)
            continue
        stripped = p.mount_path
        out_path = stripped.removesuffix(".gz") if stripped.endswith(
            ".gz") else stripped + ".out"
        await write_bytes(mounted_path(p, out_path), data)
        writes[out_path] = data
        if not keep:
            await unlink(p)
    return b"".join(stdout) or None, IOResult(writes=writes,
                                              exit_code=1 if errors else 0,
                                              stderr="".join(errors).encode()
                                              or None)


__all__ = ["gunzip"]


@dataclass(frozen=True, slots=True)
class GunzipFlags:
    keep: bool = False
    force: bool = False
    to_stdout: bool = False
    test_only: bool = False


def parse_flags(flags: Mapping[str, FlagValue]) -> GunzipFlags:
    fl = FlagView(flags, spec=SPECS["gunzip"])
    return GunzipFlags(
        keep=fl.as_bool("k"),
        force=fl.as_bool("f"),
        to_stdout=fl.as_bool("c"),
        test_only=fl.as_bool("t"),
    )


def gunzip_writes(flags: Mapping[str, FlagValue],
                  paths: list[PathSpec]) -> bool:
    """Whether a gunzip invocation writes: each file operand is replaced
    by its content unless ``-c`` sends it to stdout or ``-t`` only tests
    it, while a ``-`` operand, like no operand, filters stdin to stdout.

    Args:
        flags (Mapping[str, FlagValue]): the parsed flag bag.
        paths (list[PathSpec]): the operands the mount received.
    """
    parsed = parse_flags(flags)
    replaces = any(p.raw_path != "-" for p in paths)
    return replaces and not (parsed.to_stdout or parsed.test_only)


async def gunzip_generic(
    paths: list[PathSpec],
    texts: list[str],
    opts: CommandOpts,
    read_bytes: Callable[..., Awaitable[bytes]],
    write_bytes: Callable[..., Awaitable[None]],
    unlink: Callable[..., Awaitable[None]],
) -> tuple[ByteSource | None, IOResult]:
    parsed = parse_flags(opts.flags)
    return await gunzip(paths,
                        read_bytes=read_bytes,
                        write_bytes=write_bytes,
                        unlink=unlink,
                        stdin=opts.stdin,
                        keep=parsed.keep,
                        force=parsed.force,
                        to_stdout=parsed.to_stdout,
                        test_only=parsed.test_only)

File: typescript/packages/core/src/commands/builtin/generic/gunzip.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { specOf } from '../../spec/builtins.ts'
import { FlagView } from '../../spec/flag_view.ts'
import { mountedPath } from '../../../utils/key_prefix.ts'
import { IOResult, materialize } from '../../../io/types.ts'
import type { PathSpec } from '../../../types.ts'
import { gunzipChecked } from '../../../utils/compress.ts'
import { GzipDataError } from '../../../utils/errors.ts'
import type { CommandFnResult, CommandOpts, WritesFn } from '../../config.ts'
import { STDIN_OPERAND } from '../utils/constants.ts'
import { operandLabel, stdinStream } from '../utils/stream.ts'

const ENC = new TextEncoder()

function concat(chunks: Uint8Array[]): Uint8Array {
  let total = 0
  for (const c of chunks) total += c.byteLength
  const out = new Uint8Array(total)
  let offset = 0
  for (const c of chunks) {
    out.set(c, offset)
    offset += c.byteLength
  }
  return out
}

// Whether a gunzip invocation writes: each file operand is replaced by its
// content unless -c sends it to stdout or -t only tests it, while a `-`
// operand, like no operand, filters stdin to stdout. Mirrors Python's
// gunzip_writes.
export const gunzipWrites: WritesFn = (flags, paths) => {
  const fl = new FlagView(flags, specOf('gunzip'))
  const replaces = paths.some((p) => p.rawPath !== '-')
  return replaces && !(fl.asBool('c') || fl.asBool('t'))
}

export async function gunzipGeneric(
  paths: PathSpec[],
  opts: CommandOpts,
  stream: (p: PathSpec) => AsyncIterable<Uint8Array>,
  write: (p: PathSpec, data: Uint8Array) => Promise<void>,
  unlink: (p: PathSpec) => Promise<void>,
): Promise<CommandFnResult> {
  const fl = new FlagView(opts.flags, specOf('gunzip'))
  const keep = fl.asBool('k')
  const stdoutMode = fl.asBool('c')
  const testMode = fl.asBool('t')
  const read = stdinStream(stream, opts.stdin)
  const writes: Record<string, Uint8Array> = {}
  const stdout: Uint8Array[] = []
  let errors = ''
  // With no operand gunzip reads stdin. A `-` has no file to replace, so it
  // decompresses to stdout; gzip refuses to follow /dev/stdin in place, so
  // only `-` does this. An input with no gzip header is reported and skipped;
  // a truncated or corrupt one ends the run.
  for (const p of paths.length > 0 ? paths : [STDIN_OPERAND]) {
    const inPlace = !(stdoutMode || testMode || p.rawPath === '-')
    const raw = await materialize(inPlace ? stream(p) : read(p))
    let data: Uint8Array
    try {
      data = await gunzipChecked(raw)
    } catch (err) {
      if (!(err instanceof GzipDataError)) throw err
      errors += `gunzip: ${operandLabel(p, 'stdin')}: ${err.message}\n`
      if (err.fatal) break
      continue
    }
    if (testMode) continue
    if (!inPlace) {
      stdout.push(data)
      continue
    }
    const pStripped = p.mountPath
    const outPath = pStripped.endsWith('.gz') ? pStripped.slice(0, -3) : pStripped + '.out'
    await write(mountedPath(p, outPath), data)
    writes[outPath] = data
    if (!keep) await unlink(p)
  }
  return [
    stdout.length > 0 ? concat(stdout) : null,
    new IOResult({
      writes,
      exitCode: errors === '' ? 0 : 1,
      stderr: errors === '' ? null : ENC.encode(errors),
    }),
  ]
}
